src/subs.jl

```diff
@@ -9,23 +9,69 @@ end
+    # - If a single substitution rule determines the value of a real variable, just substitute it.
+    # - If a single substitution rule determines the value of a complex variable or its conjugate, substitute the appropriate
+    #   value whereever something related to this variable is found (i.e., the complex variable, the conjugate variable, or
+    #   its real or imaginary part)
+    # - If a single substitution rule determines the value of the real or imaginary part of a complex variable alone, then only
+    #   replace the real or imaginary parts if they occur explicitly. Don't do a partial substitution, i.e., `z` with the rule
+    #   `zᵣ => 1` is left alone and not changed into `1 + im*zᵢ`. Even if both the real and imaginary parts are substituted as
+    #   two individual rules (which we don't know of in this method), `z` will not be replaced.
+    if s.first.kind == REAL
+        for j in eachindex(vars)
+            if vars[j] == s.first
+                vals[j] = s.second
+                C == Commutative && break
+            end
+        end
+    else
+        for j in eachindex(vars)
+            if vars[j].variable_order.order.id ==
+               s.first.variable_order.order.id
+                if s.first.kind == COMPLEX || s.first.kind == CONJ
+                    value = s.first.kind == CONJ ? conj(s.second) : s.second
+                    if vars[j].kind == COMPLEX
+                        vals[j] = value
+                    elseif vars[j].kind == CONJ
+                        vals[j] = conj(value)
+                    elseif vars[j].kind == REAL_PART
+                        vals[j] = real(value)
+                    else
+                        vals[j] = imag(value)
+                    end
+                elseif s.first.kind == REAL_PART
+                    isreal(s.second) || error(
+                        "Cannot assign a complex value to the real part of an expression",
+                    )
+                    value = real(s.second) # just to make sure the type is correct
+                    if vars[j].kind == REAL_PART
+                        vals[j] = value
+                    elseif vars[j].kind != IMAG_PART
+                        error(
+                            "Found complex variable with substitution of real part - not implemented",
+                        )
+                    end
+                else
+                    @assert(s.first.kind == IMAG_PART)
+                    isreal(s.second) || error(
+                        "Cannot assign a complex value to the imaginary part of an expression",
+                    )
+                    value = real(s.second) # just to make sure the type is correct
+                    if vars[j].kind == IMAG_PART
+                        vals[j] = value
+                    elseif vars[j].kind != REAL_PART
+                        error(
+                            "Found complex variable with substitution of imaginary part - not implemented",
```

Comment:
Nice, could you add some tests for these edge cases ?

Reply:
I added a few testcases. Note that I did not use the substitution syntax without explicit variable specification. This is because `variables(p)` would list all variables, complex, conjugates, real, and imaginary parts separately. So you'd have to specify values for all of them (unnecessarily - and this would even be more restrictive as then: `z + real(z)`, called with arguments `(1+2im, 1)` would trigger the new error, because `z` occurs but a value is assigned to `real(z)`). Either the whole variable assignment would need to be redesigned or we just say that for complex variables, you should always explicitly specify a substitution rule.
(But even then, substitution rules could be redundant; `[z, real(z)] => [1+2im, 1]` is also a valid rule. Should we check whether the rules are self-consistent? Given that variables occuring twice are also not checked, I think the current lax behavior is fine.

Reply:
Indeed, that's an issue. We should only allow `COMPLEX_PART` to used in substitutions then. In this PR, I think we should target the simplest implementation that's bullet proof. We can try to make something more complicated but I'd rather do it in a separate PR if there is a use case for it. So I'd just restrict `z` to be a `COMPLEX_PART` whenever `z => ...` is used

Reply:
This simplifies the substitution process significantly. I don't know whether I like it, but there's some sense to it.

Reply:
We can revisit in a follow up PR without being breaking. We'll be able to recover your implementation easily from this branch.